---
title: "CRC PTPM variable comparisons"
author: "Joseph J Zhao"
date: "13/12/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
package.name=c(
  # data manipulation
  "tidyverse", "readr", "readxl", "dplyr", "tidyr", "lubridate", "tibble", "plyr", "devtools", "stringr", "stringi", "gtools", "tableone",
  # parallel processing
  "doParallel", "parallel", "foreach",
  # bioinformatics
  "DESeq2","preprocessCore", "umap",
  # survival analysis
  "survminer", "survival", "rstpm2", "survRM2",
  # regression stuff
  "MASS", "splines", "Hmisc", "rms",
  # misc
  "cluster", "ResourceSelection", "digitize", "tableone",
  # plotting
  "blandr", "pheatmap", "ggrepel", "ggplot2", "corrplot", "ggpubr", "gplots", "grid", "ggbiplot", "ggsankey", "ggforce", "ggbeeswarm", "EnvStats"
)

for (package.name in package.name){
  tryCatch(
  {
  if (!require(package.name, character.only = TRUE)){ 
    install.packages(package.name, character.only = TRUE) 
    library(package.name, character.only = TRUE, quietly = T)} else {library(package.name, character.only = TRUE, quietly = T)}
  }, 
    error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
}


```

## Set main working directory

```{r working directory}
wd="C:/Users/jzhao/OneDrive/Research_Cloud/NUH_NCIS/gccrc_pm/"
setwd(wd)

dir.create(paste(wd,"output/", Sys.Date(), sep=""))
setwd(paste(wd, "output/",Sys.Date(), sep="") )

```

## Unpaired comparisons
GC: all original data
Load count data and clinical data of combined gascad and peritoneal data
include: cohort %in% c("PIANO", "PIPAC", "GAS_MATT", "PC_JOHNNY", "PM_JOHNNY", "NCC_JOHNNY", "GAS_MATT_NORMAL", "GASCAD")

```{r}

path="C:/Users/jzhao/OneDrive/Research_Cloud/NUH_NCIS/gccrc_pm/submissions/substudies/crcpm/submissions/ccr/r1/workings"

l_gipm_rnaseq <- readRDS(paste0(path,"/data/wts/l_crc_rnaseq.rds"))
df_rna=l_gipm_rnaseq$gene_count
df_rna_norm=l_gipm_rnaseq$gene_log2fpkm

df_ref <- read_excel(paste0(path,"/data/reference/df_ref_pm_persample_annotated_crc.xlsx"))
df_ref=subset(df_ref, 
              (df_ref$rnaseq==1 | df_ref$wes==1) & 
                df_ref$tumornormal=="tumor" & 
                df_ref$gccrc=="crc" &
                df_ref$rnaseqid %in% colnames(df_rna_norm))

df_rna_norm=df_rna_norm[, colnames(df_rna_norm) %in% df_ref$rnaseqid]
df_rna=df_rna[rownames(df_rna) %in% rownames(df_rna_norm), colnames(df_rna) %in% df_ref$rnaseqid]

```

#### define groups for comparison

```{r}

ptpmlevels=c("primarytumor", "peritonealtumor")

df_ref$ptpmgroup=df_ref$sampletype
df_ref$ptpmgroup=factor(df_ref$ptpmgroup, level=ptpmlevels)

```

#### tableone comparison

```{r}

tab_all=CreateTableOne(data=subset(df_ref, select=c("emt_score", "tumorpurity", "tme_subtype","fractionalteredgenome_gatk_cnv", "wgd", "n_clones","clonality", "total_tmb_perMB_log", "dominantclone_medianmaf", "cms_subtype", "ptpmgroup")))
tab_strat=CreateTableOne(data=subset(df_ref, select=c("emt_score", "tumorpurity", "tme_subtype", "fractionalteredgenome_gatk_cnv", "wgd", "n_clones","clonality",  "total_tmb_perMB_log", "dominantclone_medianmaf","cms_subtype","ptpmgroup")), strata="ptpmgroup")

tab_pmptvars=cbind(print(tab_all, showAllLevels = T, nonnormal = "n_clones"), print(tab_strat, showAllLevels = T, nonnormal = "n_clones")[,-1])

write.csv(tab_pmptvars, "gccrcpm_tab_pmptvars.csv")

```

### continous variables

```{r}

vars=c("emt_score", "tumorpurity", "fractionalteredgenome_gatk_cnv", "fract", "dominantclone_medianmaf", "n_clones", "total_tmb_perMB_log")
vars_clean=c("emt_score"="EMT score", 
             "tumorpurity"="Tumor purity", 
             "demo_age"="Age", 
             "fractionalteredgenome_gatk_cnv"="Fraction of altered genome",
             "dominantclone_medianmaf"="Median MAF of dominant clone",
             "n_clones"="Number of clones",
             "fract"="Fraction",
             "total_tmb_perMB_log"="log(TMB per MB)"
             )

for (i.gccrc in unique(df_ref$gccrc)){
  
  pdf(paste(Sys.Date(),i.gccrc, "pmptvariablecomparisons_continous.pdf"), height=3, width=2.5)
  for (i.var in vars){
    
    df_violin=subset(df_ref, df_ref$gccrc==i.gccrc, select=c("ptpmgroup", i.var))
    colnames(df_violin)[grepl(i.var, colnames(df_violin))]="value"
    df_violin=subset(df_violin, !is.na(df_violin$value))
    
    # compare means
    comps=compare_means(value ~ ptpmgroup,  data = df_violin, method="t.test")
    comps=subset(comps, comps$group1=="peritonealtumor" | comps$group2=="peritonealtumor")
    my_comparisons=list()
    for(i in 1:nrow(comps)){my_comparisons[[i]]=as.vector(unlist(comps[i, c(2:3)]))}
  
    plot_violin=ggplot() +
                            geom_boxplot(aes(y=value, x=ptpmgroup, fill=ptpmgroup), data=subset(df_violin, !is.na(df_violin$ptpmgroup)),  width=0.25,  color="black", notch = F, notchwidth = 0.65)+
                            scale_fill_manual(values=c("primarytumor"="dodgerblue3", "peritonealtumor"="firebrick3"))+
                            geom_beeswarm(aes(y=value, x=ptpmgroup, fill=ptpmgroup), data=subset(df_violin, !is.na(df_violin$ptpmgroup)), cex=1, size=1, alpha=0.5)+
                            theme_classic()+
                            stat_compare_means(aes(y=value, x=ptpmgroup, label = sprintf("p = %5.4f", as.numeric(..p.format..))), data=subset(df_violin, !is.na(df_violin$ptpmgroup)), size=2.75, method="t.test", step.increase=0.15, tip.length=0.05)+ 
                            labs(x="", y="", title=vars_clean[[i.var]], color="") +
                            guides(fill = "none", alpha="none", color="none") +
                            stat_n_text(size=3.2)
    
    plot_violin %>% print
   
    
    }
  
  dev.off()
}



```

### categorical variables

```{r}

vars=c("cms_subtype", "tme_subtype", "wgd", "clonality")
vars_clean=c("mesgc"="MesGC classification (Ho et al.)", 
             "wangetal_gastric" = "GI tumor lineage (Wang et al.)", 
             "tme_subtype"="TME subtype (Bagaev et al.)", 
             "tumor_lauren" = "Lauren classification", 
             "tcga"="TCGA classification", 
             "cms_subtype"="CMS subtype",
             "wgd"="Whole genome duplication",
             "clonality"="Clonality")

my_colour=list(
  "cms_subtype" = c("CMS1"="slateblue3", "CMS2"="lightseagreen", "CMS3"="yellow3", "CMS4"="pink2", "primarytumor"="dodgerblue3","peritonealtumor"="firebrick3"), 
   "tme_subtype" = c("D"="steelblue4", "F"="green4", "IE"="black", "IE/F"="grey"),
  "wgd"=c("yes"="black", "no"="grey","primary-tumor-no PM"="goldenrod3", "primary-tumor-with PM"="dodgerblue3", "peritoneal-tumor"="firebrick3", "liver-tumor"="grey10"),
  "clonality"=c("monoclonal"="black", "polyclonal"="grey","primary-tumor-no PM"="goldenrod3", "primary-tumor-with PM"="dodgerblue3", "peritoneal-tumor"="firebrick3", "liver-tumor"="grey10")
)

for (i.gccrc in unique(df_ref$gccrc)[1]){
  
  pdf(paste(Sys.Date(),i.gccrc, "pmptvariablecomparisons_cat.pdf"), height=2.6, width=4)
  for (i.var in vars){
    
    df=subset(df_ref, df_ref$gccrc==i.gccrc, select=c("ptpmgroup", i.var))
    df=subset(df, !is.na(df[,2]))
    colnames(df)[grepl(i.var, colnames(df))]="value"
    df=as.data.frame.matrix(table(as.character(df$ptpmgroup), df$value))
    
    set.seed(1)
    fisher=fisher.test(df, workspace = 2e7, simulate.p.value=TRUE)
    
    df=apply(df, 1, function(x){x/sum(x)})
    df_melt=reshape2::melt(df)
    
    if(i.var=="tcga"){df_melt$Var1=factor(df_melt$Var1, levels=c("primary-tumor-no PM", "primary-tumor-with PM", "peritoneal-tumor", "liver-tumor", "MSI", "EBV", "CIN", "GS"))}
    if(i.var=="wangetal_gastric"){df_melt$Var1=factor(df_melt$Var1, levels=c("primary-tumor-no PM", "primary-tumor-with PM", "peritoneal-tumor", "liver-tumor", "Gastric-dominant", "GI-mixed"))}
    if(i.var=="wgd"){df_melt$Var1=factor(df_melt$Var1, levels=c("primary-tumor-no PM", "primary-tumor-with PM", "peritoneal-tumor", "no", "yes"))}
    
    df_melt$Var2=factor(df_melt$Var2, levels=c("primarytumor", "peritonealtumor"))
    
    plot_stack=ggplot(df_melt, aes(y=value, x=Var2, fill=Var1, alpha=Var1)) +
                            geom_hline(yintercept =c(0.25, 0.5, 0.75), alpha=0.5, linetype=2)+                      
                            geom_bar(position="stack", stat="identity", width=0.98, alpha=0.8) +
                            # scale_x_discrete(guide = guide_axis(n.dodge = 2))+ 
                            scale_y_continuous(breaks=c(0, 0.25, 0.5, 0.75, 1))+
                            theme_classic2()+
                            labs(x="", y="", title=vars_clean[[i.var]], color="", fill="", subtitle = paste0(" p = ", format(scientific(fisher$p.value)))) +
                            guides(alpha="none", color="none") +
                            scale_fill_manual(values=my_colour[[i.var]]) +
      
                            if(ncol(df)==3){
                              scale_alpha_manual(values=c(c(1,1,1,0.75), rep(0.75, nrow(df))))
                            } else {
                              scale_alpha_manual(values=c(c(1,1,1,1), rep(0.75, nrow(df))))
                              }
      
                            
                            
    
    plot_stack %>% print
   
    
    }
  
  dev.off()
}



```



