---
title: "CRCPM PTPM target gene analysis"
author: "Joseph J Zhao"
date: "4/12/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
package.name=c(
  # data manipulation
  "tidyverse", "readr", "readxl", "dplyr", "tidyr", "lubridate", "tibble", "plyr", "devtools", "stringr", "stringi", "gtools", "scales",
  # parallel processing
  "doParallel", "parallel", "foreach", "doSNOW",
  # bioinformatics
  "DESeq2", "BiocParallel", "clusterProfiler", "DGEobj.utils", "umap",
  # survival analysis
  "survminer", "survival", "rstpm2", "survRM2",
  # regression stuff
  "MASS", "splines", "Hmisc", "rms",
  # misc
  "cluster", "ResourceSelection", "digitize", "rstatix",
  # plotting
  "blandr", "pheatmap", "ggrepel", "ggplot2", "corrplot", "ggpubr", "gplots", "grid", "gridExtra", "EnvStats"
)

for (package.name in package.name){
  tryCatch(
  {
  if (!require(package.name, character.only = TRUE)){ 
    install.packages(package.name, character.only = TRUE) 
    library(package.name, character.only = TRUE)} else {library(package.name, character.only = TRUE)}
  }, 
    error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
}

```

## Set main working directory

```{r working directory}
wd="C:/Users/jzhao/OneDrive/Research_Cloud/NUH_NCIS/gccrc_pm/"
setwd(wd)

dir.create(paste(wd,"output/", Sys.Date(), sep=""))
setwd(paste(wd, "output/",Sys.Date(), sep="") )

```

load datasets

```{r}

path="C:/Users/jzhao/OneDrive/Research_Cloud/NUH_NCIS/gccrc_pm/submissions/substudies/crcpm/submissions/ccr/r1/workings"

l_gipm_rnaseq <- readRDS(paste0(path,"/data/wts/l_crc_rnaseq.rds"))
df_rna=l_gipm_rnaseq$gene_count
df_rna_norm=l_gipm_rnaseq$gene_log2fpkm

df_ref <- read_excel(paste0(path,"/data/reference/df_ref_pm_persample_annotated_crc.xlsx")) %>% data.frame
df_ref=subset(df_ref, df_ref$rnaseq==1  & !df_ref$gccrc %in% c("benign") & df_ref$tumornormal=="tumor")

df_rna=df_rna[, colnames(df_rna) %in% df_ref$rnaseqid]
df_rna_norm=df_rna_norm[, colnames(df_rna_norm) %in% df_ref$rnaseqid]

df_ref=subset(df_ref, df_ref$rnaseqid %in% colnames(df_rna_norm)); rownames(df_ref)=df_ref$rnaseqid

table(df_ref$sampletype, df_ref$gccrc)

```

# define groups for comparison

```{r}

# df_ref$ptpmgroup=paste(df_ref$primarypm, df_ref$tumornormal, df_ref$pm_onset_datebased, sep="-")
# df_ref$ptpmgroup[grepl("peritoneal-tumor", df_ref$ptpmgroup)]="peritoneal-tumor"
# df_ref$ptpmgroup[grepl("liver-tumor", df_ref$ptpmgroup)]="liver-tumor"
# df_ref$ptpmgroup %>% table
# 
# df_ref$ptpmgroup=factor(df_ref$ptpmgroup, levels=ptpmlevels)

```


## define genes of interest
https://ascopubs.org/doi/full/10.1200/EDBK_391278
genes_kreidieh=c("CTLA4", "PD1", "PDL1", "LAG3", "TIM3", "TIGIT", "PVRIG", "NKG2A", "KLRC1", "CD73", "ADORA2A", "CD3", "CD19", "PMEL", "BCMA", "TNFRSF17", "CD16", "CEA", "EPCAM", "CD155")

```{r}

genes_tumor=c("SERPINE1", "FGFR1", "FGFR2", "PDL1", "MET", "RAS", "RAF", "MEK", "BRAF", "EGFR", "MYC", "PIK3CA", "DKK1", "TROP2A", "TACSTD2", "APC", "FBXW7", "NFE2L2", "PIK3R1", "KAL1", "NOTCH1", "SMAD4", "NTRK\\d{1}", "KEAP1", "CTNNB1", "MLH1", "STK11", "TEAD\\d{1}", "HPSE", "HPSE2", "MSN")

genes_kreidieh=c("CTLA4", "PD1", "PDL1", "LAG3", "TIM3", "TIGIT", "PVRIG", "NKG2A", "KLRC1", "CD73", "ADORA2A", "CD3", "CD19", "PMEL", "BCMA", "TNFRSF17", "CD16", "CEA", "EPCAM", "CD155")
genes_immune=c("CD4A", "CD8A", "CTLA4", "FOXP3", "GZM1", "GZMB", "IFNG", "LAG3", "PDCD1", "CD274", "PRF1", "HAVCR2", "TIGIT", "ILT", "IDO\\d{1}", "PDL2")

genes_immune=c(genes_immune, genes_kreidieh) %>% unique

genes_tumor=c(rownames(df_rna)[grep(paste(paste0("^", genes_tumor, "$"), collapse="|"), rownames(df_rna))])
genes_immune=c(rownames(df_rna)[grep(paste(paste0("^", genes_immune, "$"), collapse="|"), rownames(df_rna))])

genes=c(genes_tumor, genes_immune)

```

## combine expression with reference sheet

```{r}

df_ref_exp=merge(df_ref, t(data.frame(df_rna_norm[genes,])), by="row.names")
rownames(df_ref_exp)=df_ref_exp$Row.names
df_ref_exp$Row.names=NULL

```

# Expression of gene across different clinical subgroups
## Beeswarm plots of gene expression across clincal subgroups of interest

Paired violin plots PT with PM vs  PM 

```{r}
  
df_ref_exp$gccrc[df_ref_exp$gccrc=="crc"]="Colorectal"

df_pvals=NULL

for (i.gccrc in unique(df_ref_exp$gccrc)){
    
    pdf(paste(Sys.Date(),i.gccrc, "peritoneal_targetgene_subgroupexp_violin",".pdf"), height=2.4, width=2.4)
    for (i.gene in genes){

    df_ref_exp_temp=subset(df_ref_exp, df_ref_exp$gccrc==i.gccrc & grepl("tumor", df_ref_exp$sampletype))
    
    colnames(df_ref_exp_temp)[colnames(df_ref_exp_temp)==i.gene]="gene"
    df_ref_exp_temp=subset(df_ref_exp_temp, df_ref_exp_temp$gene>(-10)) # remove outliers
    
    df_ref_exp_temp$sampletype=ifelse(df_ref_exp_temp$sampletype=="primarytumor", "PT", "PM")
    df_ref_exp_temp$sampletype=factor(df_ref_exp_temp$sampletype, levels=c("PT", "PM"))
    
    df_ref_exp_temp$patient_primarypm=paste0(df_ref_exp_temp$patientid,"@", df_ref_exp_temp$sampletype, "@", df_ref_exp_temp$gccrc)
    
    # identify paired samples
    df_ref_paired=(unlist(table(df_ref_exp_temp$patientid, df_ref_exp_temp$sampletype)))
    double_ids=df_ref_paired[df_ref_paired[,1]>1 | df_ref_paired[,2]>1, ] %>% rownames
    paired_ids=df_ref_paired[df_ref_paired[,1]>0 & df_ref_paired[,2]>0, ] %>% rownames
    
    # df for paired analysis (take mean of paired samples)
    df_ref_exp_paired=df_ref_exp_temp[df_ref_exp_temp$patientid %in% paired_ids, c("patient_primarypm", "gene")]

    df_ref_exp_paired=df_ref_exp_paired%>% group_by(patient_primarypm) %>% summarise_all(function(x){(mean(x))})
    df_ref_exp_paired$patientid=str_split(df_ref_exp_paired$patient_primarypm, "@", simplify=T)[,1]
    df_ref_exp_paired$sampletype=str_split(df_ref_exp_paired$patient_primarypm, "@", simplify=T)[,2]
    df_ref_exp_paired$gccrc=str_split(df_ref_exp_paired$patient_primarypm, "@", simplify=T)[,3]
    df_ref_exp_paired=df_ref_exp_paired[order(df_ref_exp_paired$patientid),]

    comps=compare_means(gene ~ sampletype,  data = df_ref_exp_paired, method="t.test", paired=T)
    t=t.test(subset(df_ref_exp_paired, df_ref_exp_paired$sampletype=="PM")$gene,  subset(df_ref_exp_paired, df_ref_exp_paired$sampletype=="PT")$gene, paired=T)
    
    my_comparisons=list()
    for(i in 1:nrow(comps)){my_comparisons[[i]]=as.vector(unlist(comps[i, c(2:3)]))}
    
    df_pvals=rbind(df_pvals, c(geneSymbol=i.gene, t_statistic=t$statistic[[1]], p_paired=comps$p))
    
    stat.test <- df_ref_exp_temp %>% t_test(gene ~ sampletype)
    stat.test <- stat.test %>% add_xy_position(x = "sampletype")
    stat.test$p.scient <- format(stat.test$p, scientific = TRUE)
    
    plot_violin=ggplot(data=df_ref_exp_temp, aes(y=gene, x=sampletype)) +
                            geom_violin(trim=F, aes(x=sampletype, fill=sampletype), width=0.6, alpha=0.85, color=NA)+
                            geom_line(aes(x  = sampletype, y = gene, group = patientid), data=df_ref_exp_paired, alpha=0.6) +
                            ggbeeswarm::geom_beeswarm(cex=1.9, size=0.4, alpha=0.7)+                      
                            theme(legend.position = "top")+
                            theme_minimal()+
                            labs(x="", y="", title=paste0(i.gene), color="") +
                            guides(fill = "none", alpha="none") +
                            scale_fill_manual(values=c("dodgerblue3", "firebrick3"))+
                            stat_pvalue_manual(stat.test, label = paste0("unpaired T-test, p=", "{p.scient}"), tip.length = 0, step.increase = 0.08, size=2.5) +
                            annotate("text", x=1.5, y=max(df_ref_exp_temp$gene)+1, hjust=0.5, size=2.5, label=paste0("paired T-test, p=", scientific(comps$p.adj)))      
     
    
    print(plot_violin)

    }
    dev.off();dev.off();dev.off();dev.off()
}



```

# pval plot

```{r}

df_pvals=as.data.frame(df_pvals)
df_pvals[,c(2:3)]=apply(df_pvals[,c(2:3)], 2, as.numeric)
df_pvals=subset(df_pvals, df_pvals$p_paired<0.2)
df_pvals=df_pvals[rev(order(df_pvals$p_paired)), ]
df_pvals$geneSymbol=factor(df_pvals$geneSymbol, levels=unique(df_pvals$geneSymbol))
df_pvals$diffexp=ifelse(df_pvals$t_statistic>0 & df_pvals$p_paired<0.05, "higher in PM", "higher in PT")
df_pvals$geneSymbol_lab=ifelse(df_pvals$p_paired<0.1, levels(df_pvals$geneSymbol), NA)
df_pvals$geneSymbol_col=ifelse(df_pvals$geneSymbol_lab %in% genes_immune, "immune related", "tumor related")

pdf(paste(Sys.Date(),i.gccrc, "peritoneal_targetgene_pvalplot",".pdf"), height=4, width=4.5)

ggplot()+
  geom_point(aes(x=geneSymbol, y=-log(p_paired) ,col=diffexp,  alpha=p_paired<0.05), data=df_pvals) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
        scale_y_log10() +
        scale_color_manual(values=c("firebrick3", "dodgerblue3", "seagreen", "orange2")) +
        scale_alpha_manual(values=c(0.4, 1)) +
        labs(y="-log(paired p-value)", col="")+
        guides(alpha="none")+
        geom_hline(yintercept = -log(0.05)) +
        geom_text_repel(aes(x=geneSymbol, y=-log(p_paired), label=geneSymbol_lab, col=geneSymbol_col, alpha=p_paired<0.05), data=df_pvals, point.padding = 0.15,  min.segment.length = .2, box.padding = .3, lwd = 2, max.overlaps = 30, size=3.5) 

dev.off()

```
