---
title: "CRCPM - target gene changes by stromal cluster"
author: "Joseph J Zhao"
date: "4/12/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
package.name=c(
  # data manipulation
  "tidyverse", "readr", "readxl", "dplyr", "tidyr", "lubridate", "tibble", "plyr", "devtools", "stringr", "stringi", "gtools",
  # parallel processing
  "doParallel", "parallel", "foreach", 
  # bioinformatics
  "DESeq2", "BiocParallel", "clusterProfiler", "singscore", "GSVA", "CMScaller", "EnvStats", "org.Hs.eg.db", "SpatialDecon",
  # survival analysis
  "survminer", "survival", "rstpm2", "survRM2",
  # regression stuff
  "MASS", "splines", "Hmisc", "rms",
  # misc
  "cluster", "ResourceSelection", "digitize",
  # plotting
  "blandr", "pheatmap", "ggrepel", "ggplot2", "corrplot", "ggpubr", "gplots", "grid", "scales"
)

for (package.name in package.name){
  tryCatch(
  {
  if (!require(package.name, character.only = TRUE)){ 
    install.packages(package.name, character.only = TRUE) 
    library(package.name, character.only = TRUE)} else {library(package.name, character.only = TRUE)}
  }, 
    error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
}

```

## Set main working directory

```{r working directory}
wd="C:/Users/jzhao/OneDrive/Research_Cloud/NUH_NCIS/gccrc_pm/"
setwd(wd)

dir.create(paste(wd,"output/", Sys.Date(), sep=""))
setwd(paste(wd, "output/",Sys.Date(), sep="") )

```

# tumor to stroma interaction
- evaluated at patient level
- evaluated at slide level

Patient level - already done in the clustering process

```{r}

path="C:/Users/jzhao/OneDrive/Research_Cloud/NUH_NCIS/gccrc_pm/submissions/substudies/crcpm/submissions/ccr/r1/workings"

dsp_crc <- readRDS(paste0(path,"/data/dsp/dsp_crc.rds"))

crcpm_dspstromalclusters <- read_csv(paste0(path, "/data/dsp/crcpm_stromalclusters/crcpm_dspstromalclusters_results.csv"))[-1]
df_ref_stromalclusters_perpatient <- read_csv(paste0(path, "/data/dsp/crcpm_stromalclusters/df_ref_stromalclusters_perpatient.csv"))

df_ref_pm_persample <- read_excel(paste0(path,"/data/reference/df_ref_pm_persample_annotated_crc.xlsx"))
df_ref_pm_persample=left_join(df_ref_pm_persample, subset(df_ref_stromalclusters_perpatient, select=c("patientid", grep("peritoneal", colnames(df_ref_stromalclusters_perpatient), value=T))), by="patientid")

```

Creating patient to slide id - does each patient/slide have SC2?

```{r}

pData(dsp_crc)$patientslideid=paste0(pData(dsp_crc)$patientid, "_",str_replace_all(str_extract(pData(dsp_crc)[,1], "(\\w|\\ )*"), "\\ ", ""))

df_ref_patientslideid=subset(pData(dsp_crc), select=c("dspid", "patientslideid")); rownames(df_ref_patientslideid)=NULL
df_ref_patientslideid=left_join(df_ref_patientslideid, crcpm_dspstromalclusters, by="dspid")
df_ref_patientslideid$stromal_cluster[is.na(df_ref_patientslideid$stromal_cluster)]="tumor only"

df_ref_patientslideid_tab=as.data.frame.matrix(table(df_ref_patientslideid$patientslideid, df_ref_patientslideid$stromal_cluster))

df_ref_patientslideid_tab$patientslide_stromalsubtype1=ifelse(!df_ref_patientslideid_tab$SC1==0 & (df_ref_patientslideid_tab$SC2==0 & df_ref_patientslideid_tab$SC3==0), "SC1", 
       ifelse(df_ref_patientslideid_tab$SC2+df_ref_patientslideid_tab$SC1+df_ref_patientslideid_tab$SC3==0, "tumor only", "SC2 or SC3")
       )

df_ref_patientslideid_tab$patientslide_stromalsubtype2=ifelse(!df_ref_patientslideid_tab$SC2==0 & (df_ref_patientslideid_tab$SC1==0 & df_ref_patientslideid_tab$SC3==0), "SC2", 
       ifelse(df_ref_patientslideid_tab$SC2+df_ref_patientslideid_tab$SC1+df_ref_patientslideid_tab$SC3==0, "tumor only", "SC1 or SC3")
       )

df_ref_patientslideid_tab$patientslide_stromalsubtype3=ifelse(!df_ref_patientslideid_tab$SC3==0 & (df_ref_patientslideid_tab$SC1==0 & df_ref_patientslideid_tab$SC2==0), "SC3", 
       ifelse(df_ref_patientslideid_tab$SC2+df_ref_patientslideid_tab$SC1+df_ref_patientslideid_tab$SC3==0, "tumor only", "SC1 or SC2")
       )

df_ref_patientslideid_tab$patientslideid=rownames(df_ref_patientslideid_tab)
# df_ref_patientslideid_tab$patientid=str_split(df_ref_patientslideid_tab$patientslideid, "_", simplify=T)[,1]

df_ref_patientslideid=left_join(df_ref_patientslideid, df_ref_patientslideid_tab, by="patientslideid")

# combine
df_ref_pm_persample=left_join(df_ref_pm_persample, df_ref_patientslideid, by="dspid")


```

now we want to delineate differences (primay vs peritoneum, SC2 vs non SC2)
- pathways
- immune cell types
- target genes

prepare pathway, immunecell type and target genes

```{r}

# pathways
l_pathways=readRDS(paste0(path,"/data/reference/genesets/l_gspathway.RData"))
names(l_pathways)=str_replace_all(names(l_pathways), "HALLMARK ", "")
df_dsp_pathways=gsva(dsp_crc@assayData$q_norm, l_pathways, method="gsva") %>% t %>% as.data.frame()

# target genes
genes_tumor=c("SERPINE1", "FGFR1", "FGFR2", "PDL1", "MET", "RAS", "RAF", "MEK", "BRAF", "EGFR", "MYC", "PIK3CA", "DKK1", "TROP2A", "TACSTD2", "APC", "FBXW7", "NFE2L2", "PIK3R1", "KAL1", "NOTCH1", "SMAD4", "NTRK\\d{1}", "KEAP1", "CTNNB1", "MLH1", "STK11", "TEAD\\d{1}", "HPSE", "HPSE2")

genes_kreidieh=c("CTLA4", "PD1", "PDL1", "LAG3", "TIM3", "TIGIT", "PVRIG", "NKG2A", "KLRC1", "CD73", "ADORA2A", "CD3", "CD19", "PMEL", "BCMA", "TNFRSF17", "CD16", "CEA", "EPCAM", "CD155")
genes_immune=c("CD4A", "CD8A", "CTLA4", "FOXP3", "GZM1", "GZMB", "IFNG", "LAG3", "PDCD1", "CD274", "PRF1", "HAVCR2", "TIGIT", "ILT", "IDO\\d{1}", "PDL2")
genes_immune=c(genes_immune, genes_kreidieh) %>% unique

genes_tumor=c(rownames(dsp_crc@assayData$q_norm)[grep(paste(paste0("^", genes_tumor, "$"), collapse="|"), rownames(dsp_crc@assayData$q_norm))])
genes_immune=c(rownames(dsp_crc@assayData$q_norm)[grep(paste(paste0("^", genes_immune, "$"), collapse="|"), rownames(dsp_crc@assayData$q_norm))])

genes=c(genes_tumor, genes_immune)

df_dsp_targetgene=t(dsp_crc@assayData$q_norm[grep(paste(paste0("^", genes, "$"), collapse="|"), rownames(dsp_crc@assayData$q_norm)), ]) %>%  as.data.frame()

# immune decon
data("safeTME"); head(safeTME)
data("safeTME.matches"); head(safeTME.matches)

restils = runspatialdecon(object = dsp_crc,
                          norm_elt = "q_norm",                    # normalized data
                          raw_elt = "exprs_batch",                      # expected background counts for every data point in norm
                          X = safeTME,                            # safeTME matrix, used by default
                          cellmerges = safeTME.matches,           # safeTME.matches object, used by default
                          cell_counts = dsp_crc$nuclei,           # nuclei counts, used to estimate total cells
                          is_pure_tumor = grepl("tumor", pData(dsp_crc)$group),          # identities of the Tumor segments/observations
                          n_tumor_clusters = 5)

df_dsp_spatialdecon=restils$prop_of_all %>% as.data.frame
rownames(df_dsp_spatialdecon)=str_replace_all(str_replace_all(rownames(df_dsp_spatialdecon), ".dcc", ""), "-", "_")
colnames(df_dsp_spatialdecon)=str_replace_all(colnames(df_dsp_spatialdecon), "\\.", " ")



```

list them all

```{r}

l_dspderivatives=list(
  pathways=df_dsp_pathways,  
  targetgene=df_dsp_targetgene
  # immunecell=df_dsp_spatialdecon
)

```

create loop for comparisons

```{r}


for (i.dspderivatives in names(l_dspderivatives)[2]){
  
  df_dspderivatives=l_dspderivatives[[i.dspderivatives]]
  
      # df_temp=subset(df_ref_pm_persample, df_ref_pm_persample$gccrc=="crc")
      df_dspderivatives$dspid=rownames(df_dspderivatives)
      
      df_temp=left_join(crcpm_dspstromalclusters, df_dspderivatives, by="dspid")
      df_temp$stromal_cluster[is.na(df_temp$stromal_cluster)]="tumor"
      
      df_temp=reshape2::melt(df_temp)
      
      df_temp_agg=aggregate(value~variable+stromal_cluster, data=df_temp, mean)
      df_temp_agg_cast=reshape2::dcast(variable~stromal_cluster, data=df_temp_agg) %>% as.data.frame
      rownames(df_temp_agg_cast)=df_temp_agg_cast$variable; df_temp_agg_cast$variable=NULL
      
      pdf(paste(Sys.Date(),i.dspderivatives, "crcpm_dsp_sc2heatmap.pdf"), height=8, width=5)
      sc_heatmap=pheatmap(df_temp_agg_cast, 
               scale="row",
               border_color = NA)
      
      print(sc_heatmap)
      
      
      
      dev.off()
      
      pdf(paste(Sys.Date(),i.dspderivatives, "crcpm_dsp_sc2targetgene.pdf"), height=4, width=4)

      plot_box_cell=ggplot(aes(x=variable, y=value, fill=stromal_cluster), data=subset(df_temp, df_temp$variable=="SERPINE1")) +
        labs(title = "Target gene comparisons across compartments", y="quantile normalized\ngene expression", x="", fill="") +
          geom_boxplot(width=0.6, outlier.size=0.5)+
        scale_y_log10()+
        theme_classic2()+
        stat_compare_means(method="anova", label = "p", label.y = 2.2)+
        scale_fill_manual(values=c("SC1"="goldenrod3", "SC2"="seagreen4", "SC3"="pink3", "tumor"="steelblue3"))
      
      print(plot_box_cell)
      
      dev.off()
  
}

dev.off();dev.off();dev.off();dev.off()

```


comparing SERPINE1 to pathways

```{r}

df_dspderivatives=l_dspderivatives[[2]]
df_dspderivatives$dspid=rownames(df_dspderivatives)

df_serpine1comp=left_join(crcpm_dspstromalclusters, subset(df_dspderivatives, select=c("SERPINE1", "dspid")), by="dspid")

df_dspderivatives=l_dspderivatives[[1]]
df_dspderivatives$dspid=rownames(df_dspderivatives)

df_serpine1comp=left_join(df_serpine1comp, df_dspderivatives, by="dspid")

for (i.sc in c("SC1", "SC2", "SC3")){
  cor.test(log10(subset(df_serpine1comp, df_serpine1comp$stromal_cluster==i.sc)$SERPINE1), subset(df_serpine1comp, df_serpine1comp$stromal_cluster==i.sc)$`EPITHELIAL MESENCHYMAL TRANSITION`)  %>% print
}

cor=cor.test(log10(subset(df_serpine1comp, !is.na(df_serpine1comp$stromal_cluster))$SERPINE1), subset(df_serpine1comp, !is.na(df_serpine1comp$stromal_cluster))$`EPITHELIAL MESENCHYMAL TRANSITION`)

plot_serpine1comp=ggplot() +
    geom_point(aes(x=SERPINE1, y=`EPITHELIAL MESENCHYMAL TRANSITION`, color=stromal_cluster), data=subset(df_serpine1comp, !is.na(df_serpine1comp$stromal_cluster)), size=2) +
    scale_x_log10()+
    labs(x="SERPINE1,\nquantile normalized gene expression", y="EMT, GSVA score") +
    scale_color_manual(values=c("SC1"="goldenrod3", "SC2"="seagreen4", "SC3"="pink3")) + 
    guides(color="none") + 
    annotate("text", x=30, y=-0.25, hjust=0, size=3, label=paste0("r=",round(cor$estimate[[1]], 3), ", p=", scientific(cor$p.value))) +
    geom_smooth(aes(x=SERPINE1, y=`EPITHELIAL MESENCHYMAL TRANSITION`), data=subset(df_serpine1comp, !is.na(df_serpine1comp$stromal_cluster)), method = 'lm', se = T, formula = y ~ x, color="black")

pdf(paste(Sys.Date(),i.dspderivatives, "crcpm_dsp_serpine1comp.pdf"), height=3, width=4)
print(plot_serpine1comp)
dev.off()


```

