---
title: "CRCPM identify top genes per sc"
author: "Joseph J Zhao"
date: "4/12/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
package.name=c(
  # data manipulation
  "tidyverse", "readr", "readxl", "dplyr", "tidyr", "lubridate", "tibble", "plyr", "devtools", "stringr", "stringi", "gtools",
  # parallel processing
  "doParallel", "parallel", "foreach", 
  # bioinformatics
  "DESeq2", "BiocParallel", "clusterProfiler", "singscore", "GSVA", "CMScaller", "EnvStats", "CancerSubtypes",
  # survival analysis
  "survminer", "survival", "rstpm2", "survRM2",
  # regression stuff
  "MASS", "splines", "Hmisc", "rms",
  # misc
  "cluster", "ResourceSelection", "digitize",
  # plotting
  "blandr", "pheatmap", "ggrepel", "ggplot2", "corrplot", "ggpubr", "gplots", "grid"
)

for (package.name in package.name){
  tryCatch(
  {
  if (!require(package.name, character.only = TRUE)){ 
    install.packages(package.name, character.only = TRUE) 
    library(package.name, character.only = TRUE)} else {library(package.name, character.only = TRUE)}
  }, 
    error=function(e){cat("ERROR :",conditionMessage(e), "\n")})
}

```

## Set main working directory

```{r working directory}

wd="C:/Users/jzhao/OneDrive/Research_Cloud/NUH_NCIS/gccrc_pm/"
setwd(wd)

dir.create(paste(wd,"output/", Sys.Date(), sep=""))
setwd(paste(wd, "output/",Sys.Date(), sep="") )

```


```{r}

path="C:/Users/jzhao/OneDrive/Research_Cloud/NUH_NCIS/gccrc_pm/submissions/substudies/crcpm/submissions/ccr/r1/workings"

l_crcpm_otherdatasets <- readRDS(paste0(path, "/data/validation_cohorts/l_crcpm_otherdatasets.rds"))

# add dsp data for internal validation
dsp_crc=readRDS(paste0(path, "/data/dsp/dsp_crc.rds"))

crcpm_dspstromalclusters <- read_csv(paste0(path, "/data/dsp/crcpm_stromalclusters/crcpm_dspstromalclusters_results.csv"))[-1]
pData(dsp_crc)=left_join(pData(dsp_crc), crcpm_dspstromalclusters, by="dspid")

# incorporate dsp dataset for internal validaiton
l_crcpm_otherdatasets$dsp_crc$ref_persample=subset(pData(dsp_crc), !is.na(pData(dsp_crc)$stromal_cluster))
l_crcpm_otherdatasets$dsp_crc$gene_log2exp=dsp_crc@assayData$q_norm[, colnames(dsp_crc@assayData$q_norm) %in% l_crcpm_otherdatasets$dsp_crc$ref_persample$dspid]
l_crcpm_otherdatasets$dsp_crc$ref_persample$rnaseqid=l_crcpm_otherdatasets$dsp_crc$ref_persample$dspid


```

identify high var genes

```{r}

topvargenes=CancerSubtypes::FSbyVar(l_crcpm_otherdatasets$dsp_crc$gene_log2exp, cut.type="topk",value=1000) %>% rownames # identify genes with high variance - use CancerSubtypes package

df_gs_up=NULL; df_gs_dn=NULL

for(i.gene in topvargenes){
  
  df_gene_exp=data.frame(gene_exp=dsp_crc@assayData$q_norm[rownames(dsp_crc@assayData$q_norm)==i.gene ,])
  df_gene_exp$dspid=rownames(df_gene_exp)
  df_gene_exp=left_join(df_gene_exp, crcpm_dspstromalclusters, by="dspid")
  df_gene_exp$stromal_cluster[is.na(df_gene_exp$stromal_cluster)]="tumor"
  df_gene_exp$stromal_cluster=factor(df_gene_exp$stromal_cluster, levels = c("SC1","tumor","SC3","SC2"))
  
  comps=compare_means(gene_exp~stromal_cluster, df_gene_exp, method = "t.test")
  comps=subset(comps, comps$group2=="SC2")
  
  comps$t_statistic=NA
  
  for (i.comp in 1:nrow(comps)){
    
    df_temp2=subset(df_gene_exp, df_gene_exp$stromal_cluster %in% c(comps$group1[i.comp], comps$group2[i.comp]))
    df_temp2$stromal_cluster=factor(df_temp2$stromal_cluster, levels=c(comps$group1[i.comp], comps$group2[i.comp]))
    
    t=t.test(gene_exp ~ stromal_cluster, data=df_temp2)
    comps$t_statistic[i.comp]=t$statistic[[1]]
    
  }
  
  if(all(comps$t_statistic>0) & all(comps$p.adj<0.05)){
    df_gs_dn=rbind(df_gs_dn, c(geneSymbol=i.gene, 
                               t_1v2=comps$t_statistic[1], 
                               padj_1v2=comps$p.adj[1], 
                               t_tumorv2=comps$t_statistic[2], 
                               padj_tumorv2=comps$p.adj[2], 
                               t_3v2=comps$t_statistic[3], 
                               padj_3v2=comps$p.adj[3]))
    
  } else if (all(comps$t_statistic<0) & all(comps$p.adj<0.05)){
    df_gs_up=rbind(df_gs_up, c(geneSymbol=i.gene, 
                               t_1v2=comps$t_statistic[1], 
                               padj_1v2=comps$p.adj[1], 
                               t_tumorv2=comps$t_statistic[2], 
                               padj_tumorv2=comps$p.adj[2], 
                               t_3v2=comps$t_statistic[3], 
                               padj_3v2=comps$p.adj[3]))
  } else {next()}
  
}

df_gs_dn=as.data.frame(df_gs_dn); df_gs_up=as.data.frame(df_gs_up)
df_gs_dn[,2:7]=apply(df_gs_dn[,2:7], 2, as.numeric)
df_gs_up[,2:7]=apply(df_gs_up[,2:7], 2, as.numeric)

df_gs_up$t_v2sum=df_gs_up$t_1v2+df_gs_up$t_3v2+df_gs_up$t_tumorv2
df_gs_dn$t_v2sum=df_gs_dn$t_1v2+df_gs_dn$t_3v2+df_gs_dn$t_tumorv2

df_gs_up=df_gs_up[order(df_gs_up$t_v2sum), ]
df_gs_dn=df_gs_dn[rev(order(df_gs_dn$t_v2sum)), ]

l_dspdegttest=list(up=df_gs_up, dn=df_gs_dn)

```

export

```{r}

saveRDS(l_dspdegttest, paste0(path, "/data/dsp/crcpm_stromalclusters/l_dsp_sc_deg.rds"))

```

